File: DFS, BFS/prog_타겟넘버.cpp
#include <string>
#include <vector>

using namespace std;

int result =0;
vector<bool> visited;
vector<int> nums;
void addNumber(int index, int target) {

    if(index >nums.size()) {
        return;
    }

    int num=0;
    for(int i=0;i <nums.size(); i++) {
        num += nums[i];
    }
    if(num == target) {
        result++;
    }

    for(int i = index; i<nums.size();i++) {
        if(visited[i]) {
            continue;
        }
        visited[i] = true;
        nums[i] = -nums[i];
        addNumber(i+1, target);
        visited[i] = false;
        nums[i] = -nums[i];
    }
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    visited.assign(numbers.size(), 0);
    nums = numbers;

    for(int i=0; i<numbers.size(); i++) {
        nums[i] = -nums[i];
        addNumber(i+1,target);
        nums[i] = -nums[i];
    }

    answer = result;
    return answer;
}

File: Java/Prog_타겟넘버.java
public class Prog_타겟넘버 {

    class Solution {
        int[] number;
    
        int size;
        int ans =0;
        
        public void dfs(int target, int index, int[] num) {
            if(index==size) {
                int result=0;
                for(int i=0; i<size; i++) {
                    result += num[i];
                }
                if(result == target) {
                    ans = ans +1;
                }
                return;
            }
        
            num[index] = number[index];
            dfs(target, index+1,num);
            
            num[index] = -number[index];
            dfs(target, index+1,num);
    
        }
        
        
        public int solution(int[] numbers, int target) {
            int answer = 0;
            number = numbers;
            size = numbers.length;
    
            int[] nums = new int[size];
            
            dfs(target, 0, nums);
            answer = ans;
            return answer;
        }
    }
}
